### PR TITLE
Feature/support for null value handling in sql view

### PR DESCRIPTION
**Pull request by yomiBazuaye**
Hi guys,
we are going to be using the initial SQL view generated by Transformer as the starting point for all of our curation. The source data is JSON so it's very likely that it will contain NULL values. 
Our curation processes will involve lots of Joins between tables, and our curated data will be consumed by many reports/users: if we handle NULL values in a consistent way we can make joins between tables more predictable and we can make life easier for everyone that is downstream from the initial view.

This PR includes:
- creation of NULL value handling logic for strings and numeric values
- creation of a "no_null_values" SQL view (screenshot attached)
- _upgraded_ testing framework from unittest to pytest
- unit tests for new functionality
image

### CODE REVIEW

**Comment by thoean** [comment]
Thank you for submitting this change. There are some public holidays and many people are out today and on Friday... it'll likely be early next week until we get to review your changes.

Thanks again for your support adding features to this library.

**Comment by mzas** [comment]
Hi @yomiBazuaye @thoean 
@yomiBazuaye thank you for your contribution. I will review the proposition until 20.08.2019.
Michał

File: j2v/generation/processor.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ def process_json_files(self, json_file_list):
         self.looker_writer.create_view_file(self.generator.views_dimensions_expr)
         self.looker_writer.create_explore_file(self.generator.explore_joins)
         self.sql_writer.print_sql(self.generator.all_fields, self.generator.all_joins)
+        self.sql_writer.print_sql_no_nulls(self.generator.all_non_null_fields, self.generator.all_joins)
```
Comment (mzas) [inline j2v/generation/processor.py]: Let's print SQL only 1 time in 1 version. We should be able to control if we want to print / return it with IFNULL checks or without them, that would simplify a code.

File: j2v/tests/test_generator.py [hunk 2]
```diff
@@ -1,52 +1,80 @@
 import os
 import sys
-import unittest
+import pytest
```
Comment (mzas) [inline j2v/tests/test_generator.py]: Thanks a lot for this change. That's useful.

File: j2v/utils/helpers.py [hunk 3]
```diff
@@ -17,9 +17,12 @@ def get_dimension_types(dim_val):
     elif type(dim_val) == bool:
         dim_type = "yesno"
         json_type = "boolean"
-    elif type(dim_val) == int:
+    elif type(dim_val) == int or type(dim_val) == float:
         dim_type = "number"
         json_type = "number"
+    elif dim_val is None:
```
Comment (mzas) [inline j2v/utils/helpers.py]: That's very useful.

File: j2v/str_templates/sql_templates.py [hunk 4]
```diff
@@ -1,2 +1,4 @@
 join_str_template = "LATERAL FLATTEN(OUTER => TRUE, INPUT => {exploded_structure_path}) {alias}"
 field_str_template = "{TABLE}.{__path}::{json_type} AS {path_alias}"
+non_nullable_text_field_str_template = "IFNULL({TABLE}.{__path}::{json_type},'N/A') AS {path_alias}"
```
Comment (mzas) [inline j2v/str_templates/sql_templates.py]: I think we should have a generic template for null fields, and have <expr2> ('N/A' OR 0) as parameters to the program. I can imagine many other popular replacements for NULL strings, such as - or a space bar. The same for numeric values.

File: j2v/generation/generator.py [hunk 5]
```diff
@@ -153,5 +155,20 @@ def __add_dimension(self, field_path_sql, current_view, dimension_name, dim_val,
         self.views_dimensions_expr[current_view].add(new_dimension)
 
         sql_select = st.field_str_template.format(__path=field_path_sql, TABLE=current_view,
-                                                  json_type=json_type, path_alias=full_path_nice.upper())
+                                              json_type=json_type, path_alias=full_path_nice.upper())
         self.all_fields[current_view].add(sql_select)
+
+        if json_type.lower() == "number":
```
Comment (mzas) [inline j2v/generation/generator.py]: `json_type` it's always lower

File: j2v/generation/generator.py [hunk 6]
```diff
@@ -153,5 +155,20 @@ def __add_dimension(self, field_path_sql, current_view, dimension_name, dim_val,
...
+            sql_no_nulls_select = st.non_nullable_numeric_field_str_template.format(__path=field_path_sql,
```
Comment (mzas) [inline j2v/generation/generator.py]: Maybe we can simplify this logic by having just 1 string template and a map with a mapping of (json_type, dim_type) to a default value for NULLs? Then we can expand it easily in the future for other types.

**Pull request closed without merging**